lib/resources/grub_conf.rb

```diff
@@ -71,35 +73,62 @@ def to_s
   ######################################################################
 
   def grub2_parse_kernel_lines(content, conf)
-    # Find all "menuentry" lines and then parse them into arrays
-    menu_entry = 0
+    menu_entries = extract_menu_entries(content)
+
+    if @kernel == 'default'
+      default_menu_entry(menu_entries, conf['GRUB_DEFAULT'])
+    else
+      menu_entries.select { |entry| entry['name'] == @kernel }[0]
+    end
+  end
+
+  def extract_menu_entries(content)
+    menu_entries = []
+
     lines = content.split("\n")
-    kernel_opts = {}
-    kernel_opts['insmod'] = []
-    lines.each_with_index do |file_line, index|
-      next unless file_line =~ /(^|\s)menuentry\s.*/
-      lines.drop(index+1).each do |kernel_line|
-        next if kernel_line =~ /(^|\s)(menu|}).*/
-        if menu_entry == conf['GRUB_DEFAULT'].to_i && @kernel == 'default'
-          if kernel_line =~ /(^|\s)initrd.*/
-            kernel_opts['initrd'] = kernel_line.split(' ')[1]
-          end
-          if kernel_line =~ /(^|\s)linux.*/
-            kernel_opts['kernel'] = kernel_line.split
-          end
-          if kernel_line =~ /(^|\s)set root=.*/
-            kernel_opts['root'] = kernel_line.split('=')[1].tr('\'', '')
-          end
-          if kernel_line =~ /(^|\s)insmod.*/
-            kernel_opts['insmod'].push(kernel_line.split(' ')[1])
-          end
-        else
-          menu_entry += 1
-          break
+    lines.each_with_index do |line, index|
+      next unless line =~ /^menuentry\s+.*/
+      entry = {}
+      entry['insmod'] = []
+
+      # Extract name from menuentry line
+      capture_data = line.match(/(?:^|\s+).*menuentry\s*['|"](.*)['|"]\s*--/)
+      entry['name'] = capture_data.captures[0]
```

Comment:
What if `.match` doesn't match the line, and then `capture_data` is `nil`?

Reply:
We should be fine since we don't make it to that bit unless `line` contains `menuentry`. That being said, I'm all for adding an exception to ensure the user gets a better error than a `nil` error.

Reply:
It's possible that a line can contain `menuentry` but still not pass your regex... unless you're willing to tell me you'd bet 3 months' salary that EVERY line in that file that has the word "menuentry" in it will ALWAYS match that regex. :)

Reply:
Hah! Solid point. It would have to contain menuentry at the start of the line followed by at least 1 whitespace. I am not willing to take that bet though!

Let me know if failing the resource is appropriate here.